crl/validator.go

```diff
@@ -20,346 +20,388 @@ package crl
+	for _, endpoint := range endpoints {
+		if crl, ok := v.getCRL(endpoint); ok {
+			if strings.Compare(crl.issuer.Subject.String(), certIssuer.Subject.String()) != 0 {
+				// We assume that an endpoint can only issue CRLs from a single CA because:
+				// If an endpoint hosts multiple CRLs, how would the server know what CRL to present?
+				// Endpoint reuse by CAs is not an issue. CAs host the CRL for some time after the CA has expired, immediate reuse result in the previous point.
+				return fmt.Errorf("multiple issuers known for CRL distribution endpoint=%s, issuers=%s,%s", endpoints, crl.issuer.Subject.String(), certIssuer.Subject.String())
+			}
+			// already exists
+			continue
 		}
+		// TODO: Optimize by starting Go routine per endpoint to update the CRL. A Go routine per CRL prevents all CRLs being updated simultaneously.
+		v.crls.Store(endpoint, newRevocationList(certIssuer))
 	}
-	return result
+	return nil
 }
 
-func (v *validator) appendCertificates(certificates []*x509.Certificate) {
-	v.certificatesLock.Lock()
-	defer v.certificatesLock.Unlock()
-
-	for _, certificate := range certificates {
-		subject := certificate.Subject.String()
-
-		if _, ok := v.certificates[subject]; !ok {
-			v.certificates[subject] = certificate
-		}
+// getCRL returns the requested crls in a save way, or returns false if it does not exist
+func (v *validator) getCRL(endpoint string) (*revocationList, bool) {
+	value, ok := v.crls.Load(endpoint)
+	if !ok {
+		return nil, false
 	}
+	return value.(*revocationList), true
 }
 
-func (v *validator) VerifyPeerCertificateFunction(maxValidityDays int) func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
-	return func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
-		// Import unknown certificates
-		var verifiedCerts []*x509.Certificate
-
-		for _, chain := range verifiedChains {
-			for _, verifiedCert := range chain {
-				if verifiedCert.IsCA {
-					verifiedCerts = append(verifiedCerts, verifiedCert)
-				}
-			}
-		}
-
-		v.appendCertificates(verifiedCerts)
-
-		// Parse certificates and check if they are revoked
-		var raw []byte
-
-		for _, rawCert := range rawCerts {
-			raw = append(raw, rawCert...)
+// sync tries to update all crls
+func (v *validator) sync() {
+	wg := &sync.WaitGroup{}
+	// Update in parallel if at least one of the issuers is still valid
+	v.crls.Range(func(endpoint2, current2 any) bool {
+		endpoint, isString := endpoint2.(string)
+		current, isCRL := current2.(*revocationList)
+		if !isString || !isCRL {
+			// should never happen
+			log.Logger().
+				WithField("endpoint", fmt.Sprintf("%v", endpoint2)).
+				WithField("CRL", fmt.Sprintf("%v", current2)).
+				Error("crl validator is invalid")
+			// TODO: should this just panic?
+			return true
 		}
-
-		certificates, err := x509.ParseCertificates(raw)
-		if err != nil {
-			return err
+		if nowFunc().Before(current.issuer.NotBefore) || nowFunc().After(current.issuer.NotAfter) {
```

Comment:
"before" is a bit obscure, theoretically the certificate might not be valid yet (and become valid in future). But I very highly doubt it's ever used that way in practice